lib/libmbedtls/core/tee_lmd_provider.c

```diff
@@ -600,92 +668,681 @@ void crypto_bignum_clear(struct bignum *s)
 {
+	if (!s)
+		return;
+	crypto_bignum_free(s->n);
+	crypto_bignum_free(s->e);
 }
 
-TEE_Result crypto_acipher_gen_rsa_key(struct rsa_keypair *key __unused,
-				      size_t key_size __unused)
+TEE_Result crypto_acipher_gen_rsa_key(struct rsa_keypair *key, size_t key_size)
 {
-	return TEE_ERROR_NOT_IMPLEMENTED;
+	TEE_Result res;
+	mbedtls_rsa_context rsa;
+	mbedtls_ctr_drbg_context ctr_drbg;
+	int lmd_res;
+	uint32_t e;
+
+	mbedtls_rsa_init(&rsa, 0, 0);
+	mbedtls_ctr_drbg_init(&ctr_drbg);
+	lmd_res = mbedtls_ctr_drbg_seed(&ctr_drbg, mbd_rand, NULL, NULL, 0);
+	if (lmd_res != 0) {
+		EMSG("mbedtls_ctr_drbg_seed ret is 0x%x", -lmd_res);
+		mbedtls_ctr_drbg_free(&ctr_drbg);
+		return TEE_ERROR_SECURITY;
+	}
+
+	/* get the public exponent */
+	mbedtls_mpi_write_binary((mbedtls_mpi *)key->e,
+				(unsigned char *)&e, sizeof(uint32_t));
+
+	e = TEE_U32_FROM_BIG_ENDIAN(e);
+	lmd_res = mbedtls_rsa_gen_key(&rsa, mbedtls_ctr_drbg_random,
+				      &ctr_drbg, key_size, (int)e);
+	if (lmd_res != 0) {
+		res = TEE_ERROR_BAD_PARAMETERS;
+	} else if ((size_t)mbedtls_mpi_bitlen(&rsa.N) != key_size) {
+		res = TEE_ERROR_BAD_PARAMETERS;
+	} else {
+		/* Copy the key */
+		crypto_bignum_copy(key->e, (void *)&rsa.E);
+		crypto_bignum_copy(key->d, (void *)&rsa.D);
+		crypto_bignum_copy(key->n, (void *)&rsa.N);
+		crypto_bignum_copy(key->p, (void *)&rsa.P);
+
+		crypto_bignum_copy(key->q, (void *)&rsa.Q);
+		crypto_bignum_copy(key->qp, (void *)&rsa.QP);
+		crypto_bignum_copy(key->dp, (void *)&rsa.DP);
+		crypto_bignum_copy(key->dq, (void *)&rsa.DQ);
+
+		res = TEE_SUCCESS;
+	}
+	/* Free the temporary key */
+	mbedtls_ctr_drbg_free(&ctr_drbg);
+	mbedtls_rsa_free(&rsa);
+
+	return res;
 }
 
-TEE_Result crypto_acipher_rsanopad_decrypt(struct rsa_keypair *key __unused,
-					   const uint8_t *src __unused,
-					   size_t src_len __unused,
-					   uint8_t *dst __unused,
-					   size_t *dst_len __unused)
+TEE_Result crypto_acipher_rsanopad_encrypt(struct rsa_public_key *key,
+					   const uint8_t *src, size_t src_len,
+					   uint8_t *dst, size_t *dst_len)
 {
-	return TEE_ERROR_NOT_IMPLEMENTED;
+	TEE_Result res = TEE_SUCCESS;
+	mbedtls_rsa_context rsa;
+	int lmd_res;
+	uint8_t *buf = NULL;
+	unsigned long blen, offset;
+
+	mbedtls_rsa_init(&rsa, 0, 0);
+
+	rsa.E = *(mbedtls_mpi *)key->e;
+	rsa.N = *(mbedtls_mpi *)key->n;
+
+	rsa.len = crypto_bignum_num_bytes((void *)&rsa.N);
+
+	blen = LMD_MAX_BITS_PER_VARIABLE / 8;
+	buf = malloc(blen);
+	if (!buf) {
+		res = TEE_ERROR_OUT_OF_MEMORY;
+		goto out;
+	}
+
+	memset(buf, 0, blen);
+	memcpy(buf + rsa.len - src_len, src, src_len);
+
+	lmd_res = mbedtls_rsa_public(&rsa, buf, buf);
+	switch (lmd_res) {
+	case MBEDTLS_ERR_RSA_BAD_INPUT_DATA:
+		EMSG("mbedtls_rsa_public() returned 0x%x", -lmd_res);
+		res = TEE_ERROR_BAD_PARAMETERS;
+		goto out;
+	case 0:
+		break;
+	default:
+		/* This will result in a panic */
+		EMSG("mbedtls_rsa_public() returned 0x%x", -lmd_res);
```

Comment:
FMSG, if kept at all

Reply:
There are many `EMSG` in this PR, do I need to change all of them to `FMSG`?

Reply:
All which are like this one.